### PR TITLE
Reimplemented scheduler to be global rather than sorta-kinda-per-level.

### PR DESCRIPTION
**Pull request by shans**
This simplified it somewhat. I also moved it out of stage-loader. Eventually we should be able to delete stage-loader altogether.

### CODE REVIEW

File: core/scheduler.js [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+  var finishedPromiseList = [];
+
+  return Promise.all(initPhases.map(function(phase) {
+    return phase.phase.init(function(stream) {
+      var promise = new Promise(function(resolve, reject) {
+        schedule(phases, phase.idx + 1, stream, resolve);
+      });
+      finishedPromiseList.push(promise);
+    });
+  })).then(function() { return Promise.all(finishedPromiseList).then(function() { })});
```
Comment (dstockwell) [inline core/scheduler.js]: There's an extra `then` with an empty function here.

File: core/scheduler.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  var initPhases = phases
+      .map(function(phase, idx) { return {phase: phase, idx: idx} })
+      .filter(function(phase) { return phase.phase.init !== undefined; });
+
+  var finishedPromiseList = [];
+
+  return Promise.all(initPhases.map(function(phase) {
+    return phase.phase.init(function(stream) {
+      var promise = new Promise(function(resolve, reject) {
+        schedule(phases, phase.idx + 1, stream, resolve);
```
Comment (dstockwell) [inline core/scheduler.js]: Hmm, I wonder how we can propagate back failure of the scheduled tasks.

File: core/scheduler.js (hunk 1 above)
Comment (shans) [inline core/scheduler.js]: Without this, the promise returned by startPhaseList yields a list of undefined values.

File: core/scheduler.js (hunk 2 above)
Comment (shans) [inline core/scheduler.js]: Future patch, I think. We probably just need to plumb failure through the returned promises?

File: core/scheduler.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+      task.stream = op.stream;
+    }
+    task.index++;
+    taskQueue.push(task);
+    done();
+  });
+  return true;
+}
+
+module.exports.startPhaseList = startPhaseList;
```
Comment (dstockwell) [inline core/scheduler.js]: Maybe runPhases? The promise doesn't resolve until they are finished.

File: core/scheduler.js (hunk 1 above)
Comment (dstockwell) [inline core/scheduler.js]: Ah.

**Comment by dstockwell** [comment]
lgtm, but there's still some random console.logs

File: package.json [hunk 4]
```diff
@@ -16,12 +16,12 @@
...
     "gulp": "^3.8.11",
     "ejs": "^2.3.1",
-    "graphlib-dot": "^0.6.1",
     "bluebird": "~2.9.32",
     "chai": "^2.2.0",
     "minimist": "^1.1.0",
     "expand-home-dir": "^0.0.2",
-    "split": "*"
+    "split": "*",
+    "graphlib-dot": "git://github.com/shans/graphlib-dot"
```
Comment (dstockwell) [inline package.json]: Is this meant to be in this patch?

Comment (shans) [inline package.json]: hmm, nope

**Comment by shans** [comment]
ok nao

**Pull request merged**